### PR TITLE
Review

### PR DESCRIPTION
**Pull request by kryptachio**

### CODE REVIEW

**Comment by Great-Odin** [comment]
Not super important but mainnetFjordAddress should probably be renamed since this will also be used in Goerli with Goerli contracts

**Comment by Great-Odin** [comment]
If mint phases are set manually why do we need whitelistEndDate ?

**Comment by Great-Odin** [comment]
what is the life expectations for this contract. Do these need to addresses need to be configurable?
```
    address private constant feltzine = 0x5e080D8b14c1DA5936509c2c9EF0168A19304202;
    address private constant artist = 0xb012A1bDCA34E1d0c2267bb50e6c53C8042eB4b6;
    address private constant dev = 0x52aA63A67b15e3C2F201c9422cAC1e81bD6ea847;
    ```

File: contracts/FjordDrop.sol [hunk 1]
```diff
@@ -84,26 +83,24 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
         _;
     }
-/*//////////////////////////////////////////////////////////////
+
+    /*//////////////////////////////////////////////////////////////
                         ONLY OWNER
 //////////////////////////////////////////////////////////////*/
 
     /// @notice set _time in  Unix Time Stamp to end the whitelist sale
+    //TODO - add onlyOwner
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol]: add onlyOwner

File: contracts/FjordDrop.sol [hunk 2]
```diff
@@ -34,38 +34,38 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
-        0x6f435948d9ad4cA0a73a3257743528469899ceec;
+    string public contractURI = "ipfs://QmdyYCtUsVsC5ymr7b4txQ6hXHpLXtJU7JXCDuHEJdXnRe";
+    address private mainnetFjordAddress = 0x6f435948d9ad4cA0a73a3257743528469899ceec;
     uint256 public whitelistEndDate;
     uint256 private constant PRICE_PER_WHITELIST_NFT = 0.02 ether;
     bytes32 public whiteListSaleMerkleRoot;
     uint32 private constant MAX_MINT_PER_WHITELIST_WALLET = 2;
     mapping(address => uint32) public mintPerWhitelistedWallet;
-    uint256 private PRICE_PER_PUBLIC_MINT;
+    //TODO, minor but should we default to something? otherwise we try to do a calculation on undefined
```
Comment (kryptachio) [inline contracts/FjordDrop.sol:59]: ^

File: contracts/FjordDrop.sol [hunk 3]
```diff
@@ -115,25 +113,28 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
 
-/*//////////////////////////////////////////////////////////////
+    /*//////////////////////////////////////////////////////////////
                             MINT
 //////////////////////////////////////////////////////////////*/
 
     /// @notice mint implementation interfacing w Erc721BurningErc20OnMint contract
 
     function mint() public override nonReentrant returns (uint256) {
+        //TODO might be safer to do something like `mintCounter >= TOTAL_SUPPLY`? there might be a one-off error too
```
Comment (kryptachio) [inline contracts/FjordDrop.sol:140]: ^

**Comment by Great-Odin** [comment]
For withdraw a safer option may be a onlyowner that sends X amount to Y wallet.
if for some reason this function fails, you wont be able to withdraw non of the parts

```
    function withdraw() public nonReentrant onlyOwner {
        uint256 balance = address(this).balance;
        Address.sendValue(payable(feltzine), (balance * 375) / 1000);
        Address.sendValue(payable(artist), (balance * 375) / 1000);
        Address.sendValue(payable(dev), (balance * 250) / 1000);
    }
    ```

File: contracts/FjordDrop.sol [hunk 4]
```diff
@@ -157,10 +158,8 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
         } else if (block.timestamp >= whitelistEndDate) {
             revert FJORD_WhitelistMintEnded();
         } else {
-            require(
-                totalMinted + amount <= MAX_MINT_PER_WHITELIST_WALLET,
-                "Max mint exceeded"
-            );
+            require(totalMinted + amount <= MAX_MINT_PER_WHITELIST_WALLET, "Max mint exceeded");
+            //TODO do we need to require if `mintCounter + amount` is more than total supply? otherwise we'll go over
```
Comment (kryptachio) [inline contracts/FjordDrop.sol:169]: ^

**Comment by Great-Odin** [comment]
will Stage be public? do we need to know what stage of minting the contract is at any moment?
Can you go back in stages? any problems you could see on that?

**Comment by Great-Odin** [comment]
Not important but a good practices would be to increase the counters for minting AFTER _mint is called

File: contracts/FjordDrop.sol [hunk 5]
```diff
@@ -34,38 +34,42 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
 
     uint16 public mintCounter;
     uint16 public constant TOTAL_SUPPLY = 525;
     string public customBaseURI;
-    string public contractURI =
-        "ipfs://QmdyYCtUsVsC5ymr7b4txQ6hXHpLXtJU7JXCDuHEJdXnRe";
-    address private mainnetFjordAddress =
-        0x6f435948d9ad4cA0a73a3257743528469899ceec;
+    string public contractURI = "ipfs://QmdyYCtUsVsC5ymr7b4txQ6hXHpLXtJU7JXCDuHEJdXnRe";
+    //TODO - remove mainnetFjordAddress
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:51]: nor really used anywhere so I suggest removing references

File: contracts/FjordDrop.sol [hunk 6]
```diff
@@ -34,38 +34,42 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
+    address private mainnetFjordAddress = 0x6f435948d9ad4cA0a73a3257743528469899ceec;
     uint256 public whitelistEndDate;
     uint256 private constant PRICE_PER_WHITELIST_NFT = 0.02 ether;
     bytes32 public whiteListSaleMerkleRoot;
     uint32 private constant MAX_MINT_PER_WHITELIST_WALLET = 2;
     mapping(address => uint32) public mintPerWhitelistedWallet;
+    //TODO - make this a public variable and add getter
+    // minor but should we default to something? otherwise we try to do a calculation on undefined
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol]: suggestion on being able to view currently set value publicly and setting this to a default

File: contracts/FjordDrop.sol [hunk 7]
```diff
@@ -34,38 +34,42 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
     uint256 private PRICE_PER_PUBLIC_MINT;
+    //TODO - removed ONLY_MINT_OWNER
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol]: ONLY_MINT_OWNER there might be a simpler solution, see comments made under before token transfer function

File: contracts/FjordDrop.sol [hunk 8]
```diff
@@ -34,38 +34,42 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
-    FJORD,
-    PUBLIC
-}
+        ONLY_MINT_OWNER,
+        NOT_ACTIVE,
+        WHITELIST,
+        FJORD,
+        PUBLIC
+    }
+    //TODO - make this a public variable and add getter
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:69]: suggestion on visibility and being able to read this

File: contracts/FjordDrop.sol [hunk 9]
```diff
@@ -175,9 +174,9 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
         }
     }
 
-    
     function publicMint(uint256 _amount) public payable {
-    if (mintCounter == TOTAL_SUPPLY) {
+        //TODO what if mintCounter is 500 and _amount is 1000?
```
Comment (kryptachio) [inline contracts/FjordDrop.sol:185]: ^

File: contracts/FjordDrop.sol [hunk 10]
```diff
@@ -84,26 +88,24 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
         _;
     }
-/*//////////////////////////////////////////////////////////////
+
+    /*//////////////////////////////////////////////////////////////
                         ONLY OWNER
 //////////////////////////////////////////////////////////////*/
 
     /// @notice set _time in  Unix Time Stamp to end the whitelist sale
+    //TODO - add onlyOwner and make external rather than public
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:108]: suggestion on onlyOwner

File: contracts/FjordDrop.sol [hunk 11]
```diff
@@ -112,28 +114,33 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
         customBaseURI = customBaseURI_;
     }
 
+    //TODO - remove setFjordContractAddress
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:117]: removing unused code

File: contracts/FjordDrop.sol [hunk 12]
```diff
@@ -112,28 +114,33 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
     function setFjordContractAddress(address mainnetFjordAddress_) external onlyOwner {
         mainnetFjordAddress = mainnetFjordAddress_;
     }
+
     //@notice: owner set the different states of the minting phase
     // 0 = NOT_STARTED, 1 = WHITELIST, 2 = FJORD, 3 = PUBLIC
+    //TODO - make external rather than public
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:124]: will only ever be called outside of contract

File: contracts/FjordDrop.sol [hunk 13]
```diff
@@ -150,17 +157,16 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
         payable
         isValidMerkleProof(merkleProof, whiteListSaleMerkleRoot)
     {
+        //TODO - add cap for whitelist (WHITELIST_ALLOCATION) and make sure mintCounter is less than WHITELIST_ALLOCATION
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:160]: whitelist cap

File: contracts/FjordDrop.sol [hunk 14]
```diff
@@ -175,9 +181,9 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
         }
     }
 
-    
     function publicMint(uint256 _amount) public payable {
-    if (mintCounter == TOTAL_SUPPLY) {
+        //TODO would be safer to do `mintCounter >= TOTAL_SUPPLY`
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol]: making sure total supply cant be exceeded

File: contracts/FjordDrop.sol [hunk 15]
```diff
@@ -194,35 +200,33 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
-function _beforeTokenTransfer(
+    function _beforeTokenTransfer(
         address from,
         address to,
         uint256 amount
     ) internal virtual override(Erc721BurningErc20OnMint) {
         require(stage != MintPhase.NOT_ACTIVE, "Minting is not active");
         // check if it's a mint through the Fjord's contract
         if (stage == MintPhase.FJORD) {
+            //TODO remove this require around mainnetFjordAddress as it serves no purpose
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:211]: remove unused code

File: contracts/FjordDrop.sol [hunk 16]
```diff
@@ -194,35 +200,33 @@ contract FjordDrop is Erc721BurningErc20OnMint, ReentrancyGuard, IERC2981 {
...
         // check if it's a mint through the Fjord's contract
         if (stage == MintPhase.FJORD) {
+            //TODO remove this require around mainnetFjordAddress as it serves no purpose 
             require(to == mainnetFjordAddress, "Invalid fjord address");
             Erc721BurningErc20OnMint._beforeTokenTransfer(from, to, amount);
-        } else if(stage == MintPhase.PUBLIC || stage == MintPhase.WHITELIST) {
+        } else if (stage == MintPhase.PUBLIC || stage == MintPhase.WHITELIST) {
             ERC721._beforeTokenTransfer(from, to, amount);
         }
+        // TODO add `else if (mint_counter == 0) {ERC721._beforeTokenTransfer(from, to, amount)}`
```
Comment (bad-boy-ri-ri) [inline contracts/FjordDrop.sol:217]: potential solution for the original mint of 25 NFTs

File: contracts/FjordDrop.sol (hunk 9 above)
Comment (bad-boy-ri-ri) [reply contracts/FjordDrop.sol:185]: agreed. so check should be if  `mintCounter + amount > TOTAL_SUPPLY` we revert

**Pull request closed without merging**